backend/src/main/kotlin/org/loculus/backend/service/files/S3Service.kt

```diff
@@ -0,0 +1,107 @@
+package org.loculus.backend.service.files
+
+import io.minio.GetPresignedObjectUrlArgs
+import io.minio.MinioClient
+import io.minio.SetObjectTagsArgs
+import io.minio.http.Method
+import org.loculus.backend.config.S3BucketConfig
+import org.loculus.backend.config.S3Config
+import org.springframework.stereotype.Service
+import java.util.concurrent.TimeUnit
+
+private const val PRESIGNED_URL_EXPIRY_SECONDS = 60 * 30
+
+@Service
+class S3Service(private val s3Config: S3Config) {
+    private var client: MinioClient? = null
+    private var internalClient: MinioClient? = null
+
+    fun createUrlToUploadPrivateFile(fileId: FileId): String {
+        val config = getS3BucketConfig()
+        return getClient().getPresignedObjectUrl(
+            GetPresignedObjectUrlArgs.builder()
+                .method(Method.PUT)
+                .bucket(config.bucket)
+                .`object`(getFileName(fileId))
+                .expiry(PRESIGNED_URL_EXPIRY_SECONDS, TimeUnit.SECONDS)
```

Comment:
IMO we should allow more time than this for upload. For upload I can imagine a scenario in which one requests upload URLs for e.g. 1000 files and then uploads them one by one. It might take more than 30 minutes to get to the 1000th file. In contrast for downloads, because IMO we should always use the canonical URL which generates a new presigned URL afresh this probably isn't important (although I guess it might be for resuming downloads, not sure).

Reply:
Valid point and good to revisit this number - I'm not sure what's best. I've added a note to our board about it, also tagging @chaoran-chen so he can thing about it as well.

I think maybe we can just leave it as is though - we might want to refine the whole upload interaction in general. We probably want multi-part upload and we probably also want batching. The client should maybe also request URLs in batches (would be my first intuition).

Either way, let's revisit and I'll resolve it for now.

Reply:
I'll make an issue so it doesn't become forgotten